src/main/java/org/breedinginsight/brapi/v2/services/BrAPITrialService.java

```diff
@@ -52,4 +93,278 @@ private long countGermplasm(UUID programId, String trialDbId) throws ApiExceptio
         return obUnits.stream().map(BrAPIObservationUnit::getGermplasmDbId).distinct().count();
     }
+
+    public DownloadFile exportObservations(
+            Program program,
+            UUID experimentId,
+            ExperimentExportQuery params) throws IOException, DoesNotExistException, ApiException {
+        StreamedFile downloadFile;
+        boolean isDataset = false;
+        List<BrAPIObservation> dataset = new ArrayList<>();
+        List<BrAPIObservationVariable> obsVars = new ArrayList<>();
+        Map<String, Map<String, Object>> rowByOUId = new HashMap<>();
+        Map<String, BrAPIStudy> studyByDbId = new HashMap<>();
+        List<String> requestedEnvNames = StringUtils.isNotBlank(params.getEnvironments()) ?
+                new ArrayList<>(Arrays.asList(params.getEnvironments().split(","))) : new ArrayList<>();
+        FileType fileType = params.getFileExtension();
+
+        // get requested environments for the experiment
+        List<BrAPIStudy> expStudies = studyDAO.getStudiesByExperimentID(experimentId, program);
+        if (!requestedEnvNames.isEmpty()) {
+            expStudies = expStudies
+                    .stream()
+                    .filter(study -> requestedEnvNames.contains(
+                            Utilities.removeProgramKeyAndUnknownAdditionalData(study.getStudyName(), program.getKey())))
+                    .collect(Collectors.toList());
+        }
+        expStudies.forEach(study -> studyByDbId.putIfAbsent(study.getStudyDbId(), study));
+
+        // get the OUs for the requested environments
+        List<BrAPIObservationUnit> ous = new ArrayList<>();
+        try {
+            for (BrAPIStudy study: expStudies) {
+                ous.addAll(ouDAO.getObservationUnitsForStudyDbId(study.getStudyDbId(), program));
+            }
+        } catch (ApiException err) {
+            log.error("Error fetching observation units for a study by its DbId" +
+                    Utilities.generateApiExceptionLogMessage(err), err);
+            err.printStackTrace();
+        }
+
+        // make export columns including columns for requested dataset obsvars and timestamps if requested
+        List<Column> columns = ExperimentFileColumns.getOrderedColumns();
+        BrAPITrial experiment = getExperiment(program, experimentId);
+        if ((StringUtils.isBlank(params.getDataset()) || "observations".equalsIgnoreCase(params.getDataset())) &&
+                experiment.getAdditionalInfo().getAsJsonObject().get(BrAPIAdditionalInfoFields.OBSERVATION_DATASET_ID) != null) {
+            String obsDatasetId = experiment
+                    .getAdditionalInfo().getAsJsonObject()
+                    .get(BrAPIAdditionalInfoFields.OBSERVATION_DATASET_ID).getAsString();
+            isDataset = true;
+            obsVars = getDatasetObsVars(obsDatasetId, program);
+            addObsVarColumns(columns, obsVars, params.isIncludeTimestamps(), program);
```

Comment:
It's hard to tell what is being mutated here without looking at the implementation of `addBrAPIObsToRecords`, could you add a descriptive comment above this line?

Reply:
comment added to be more explicit

Reply:
Timestamps are downloading correctly, now.

Reply:
Ok, I added a comment to clarify.